gossipsub v1.1 specification (#254)
libp2p/specs: pubsub/gossipsub/gossipsub-v1.1.md

> # gossipsub v1.1: gossipsub extensions to improve bootstrapping and attack resistance

Reviewer: ```suggestion
# gossipsub v1.1: Security extensions to improve attack resilience and bootstrapping
```

| Lifecycle Stage | Maturity       | Status | Latest Revision |
|-----------------|----------------|--------|-----------------|
| 1A              | Draft          | Active | r1, 2020-03-20  |

Authors: [@vyzo]

Interest Group: [@yusefnapora], [@raulk], [@whyrusleeping], [@Stebalien], [@daviddias]

[@whyrusleeping]: https://github.com/whyrusleeping
[@yusefnapora]: https://github.com/yusefnapora
[@raulk]: https://github.com/raulk
[@vyzo]: https://github.com/vyzo
[@Stebalien]: https://github.com/Stebalien
[@daviddias]: https://github.com/daviddias

See the [lifecycle document][lifecycle-spec] for context about maturity level
and spec status.

[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md

---

## Overview

This document specifies extensions to [gossipsub v1.0](gossipsub-v1.0.md) intended to improve
bootstrapping and protocol attack resistant. The extensions change the algorithms that
prescribe local peer behaviour and are fully backwards compatible with v1.0 of the protocol.

> Peers that implement these extensions, advertise v1.1 of the protocol using `/meshsub/1.1.0`

Reviewer: Why not `/gossipsub/1.1.0`?

Reviewer: historical artifact?

Author: well, the intention is to some day have episub as `/meshsub/2.0.0` -- it's in the same family of protocols.

as the protocol string.

> 

Reviewer: ```suggestion

# Considered Attack Vectors

`To be written`

```

It would be good to add a section that we briefly explain the Attack Vectors we considered as it will inform the reader how the extensions proposed impact them.

> ## Peer Exchange

Reviewer: ```suggestion
## Protocol extensions 

### Peer Exchange
```

set of nodes, without relying on an external peer discovery service.

Peer Exchange (PX) kicks in when pruning a mesh because of oversubscription. Instead of simply
telling the pruned peer to go away, the pruning peer _may_ provide a set of other peers where the
pruned peer can connect to reform its mesh. In addition, both the pruned and the pruning peer
add a backoff period from each other, within which they will not try to regraft. Both the pruning
and the pruned peer will immediate prune a `GRAFT` within the backoff period.
The recommended duration for the back period is 1 minute.

In order to implement PX, we extend the `PRUNE` control message to include an optional set of
peers the pruned peer can connect to. This set of includes the Peer ID and a [_signed_ peer
record](https://github.com/libp2p/specs/pull/217) for each peer exchanged.
In order to facilitate transion to the usage of signed peer records within the libp2p ecosystem,
the emitting peer is allowed to omit the signed peer record if it doesn't have one.
In this case, the pruned peer will have to utilize an external service to discover addresses for
the peer, eg the DHT.

> ### Protobuf

Reviewer: ```suggestion
#### Protobuf
```

The `ControlPrune` message is extended with a `peer` field as follows.

```protobuf
message ControlPrune {
	optional string topicID = 1;
	repeated PeerInfo peers = 2; // gossipsub v1.1 PX
}

message PeerInfo {
	optional bytes peerID = 1;
	optional bytes signedPeerRecord = 2;
}

```

> ## Flood Publishing

Reviewer: ```suggestion
### Flood Publishing
```

In gossipsub v1.0 a freshly published message is propagated through the mesh or the fanout map
if the publisher is not subscribed to the topic. In gossipsub v1.1 publishing is (optionally)

> done by publishing the message to all connected peers with a score above a publish threshold

Reviewer 2: I was mistakenly thinking that scores were meant to be communicated to other peers. I think I understand now: Each node keeps track of the scores of its "connected" nodes. Then it uses this information privately to sub-select from the scored nodes which ones to send flood messages to. (And same for adaptive gossip, below).

"Sybil" nodes are then definitionally the nodes that have low scores (i.e. low contribution within some historical window of time). And the algorithm ensures routing around them.

(see Peer Scoring below).

This behaviour is prescribed to counter eclipse attacks and ensure that a newly published message
from a honest node will reach all connected honest nodes and get out to the network at large.
When flood publishing is in use there is no point in utilizing a fanout map or emitting gossip when
the peer is a pure publisher not subscribed in the topic.

This behaviour also reduces message propagation latency as the message is injected to more points
in the network.

> ## Adaptive Gossip Dissemination

Reviewer: ```suggestion
### Adaptive Gossip Dissemination
```

In gossipsub v1.0 gossip is emitted to a fixed number of peers, as specified by the `D_lazy`
parameter. In gossipsub v1.1 the disemmination of gossip is adaptive; instead of emitting gossip
to a fixed number of peers, we emit gossip to a percentage of our peers with a minimum of `D_lazy`
peers.

The parameter controlling the emission of gossip is called the gossip _factor_. When a node wants
to emit gossip during the heartbeat, first it selects all peers with a peer score above a gossip
threshold (see Peer Scoring below). From these peers, it randomly selects gossip factor peers with
a minimum of `D_lazy`, and emits gossip to the selected peers.

The recommended value for the gossip factor is `0.25`, which with the default of 3 rounds of gossip
per message ensures that each peer has at least 50% chance of receiving gossip about a message.
More specifically, for 3 rounds of gossip, the probability of a peer _not_ receiving gossip about
a fresh message is `(3/4)³=27/64=0.421875`. So each peer receives gossip about a fresh message with
a `0.578125` probability.

This behaviour is prescribed to counter sybil attacks and ensures that a message from a honest

> node propagates in the network with high probability.

Reviewer 2: This statement does not sound 100% accurate. Sybil attacks are not entirely "countered" with high probability. For instance, if 80% of the nodes are sybils and they invest enough time/work for the scoring function to reach its decay threshold, then they will be everywhere in the network, and then they can disconnect it. I think you mean something along the lines of: Sybil attacks whose resources are below a certain limit amount (of work, or something else) will be countered with high probability. And, of course, probably outside of the scope of a spec one would want to compute the exact amount of work that suffices to take down a network of a given size.

Author: Not sure where you are getting this 11% from. Even with 90% sybils, a little over than half the honest nodes will get the message via gossip (even if the mesh breaks down completely by being taken over by sybils somehow).
This half will propagate to the other half in a few hops at most, with high probability.

> ## Peer Scoring

Reviewer: ```suggestion
### Peer Scoring
```

In gossipsub v1.1 we introduce a peer scoring component: each individual peer maintains a score
for other peers. The score is locally computed by each individual peer based on observed behaviour
and is not shared. The score is a real value, computed as weighted mix of parameters,
with pluggable application specific scoring. The score is computed across all (configured) topics
with a weighted mix, such that faulty behaviour in one topic percolates to other topics.
Furthermore, the score is retained for some period of time when a peer disconnects, so that malicious
peers cannot easily reset their score when it drops to negative and well behaving
peers don't lose their status because of a disconnection.

The intention is to detect malicious or faulty behaviour and penalize the misbehaving peers
with a negative score.

> ### Score Thresholds

Reviewer: ```suggestion
#### Score Thresholds
```

removed from the mesh. Peers with heavily negative score are further penalized or even ignored
if the score drops too low.

More specifically, the following thresholds apply:
- `0`: the baseline threshold; peers with a score below this threshold are pruned from the mesh
  during the heartbeat and ignored when looking for peers to graft. Furthermore, no PX information
  is emitted towards those peers and PX is ignored from them.
- `gossipThreshold`: when a peer's score drops below this threshold, no gossip is emitted towards
  that peer and gossip from that peer is ignored. This threshold should be negative, such that
  some information can be propagated to/from mildly negatively scoring peers.
- `publishThreshold`: when a peer's score drops below this threshold, self published messages are
  not propagated towards this peer when (flood) publishing. This threshold should be negative, and
  less than or equal to the gossip threshold.
- `graylistThreshold`: when a peer's score drops below this threshold, the peer is graylisted and
  its RPCs are ignored. This threshold must be negative, and less than the gossip/publish threshold.

> ### Heartbeat Maintenance

Reviewer: ```suggestion
#### Heartbeat Maintenance
```

The score is checked explicitly during heartbeat maintenance such that:
- Peers with negative score are pruned from all meshes.
- When pruning because of oversubscription, the peer keeps the best `D_score` scoring peers and
  selects the remaining peers to keep at random. This protects the mesh from takeover attacks
  and ensures that the best scoring peers are kept in the mesh. At the same time, we do keep some
  peers as random so that the protocol is responsive to new peers joining the mesh.
- When selecting peers to graft because of undersubscription, peers with a negative score are ignored.

> ### The Score Function

Reviewer: ```suggestion
#### The Score Function
```

  so that a peer that was pruned because of underdelivery cannot quickly get regrafted into the
  mesh. The parameter is mixed with negative weight.
- P₄: invalid messages for a topic. This is he number of invalid messages delivered in the topic.
  This is intended to penalize peers who transmit invalid messages, according to application specific
  validation rules. It is mixed with a negative weight.
- P₅: application specific score. This is the score component assigned to the peer by the application
  itself, using application specific rules. The weight is positive, but the parameter itself has an
  arbitrary real value, so that the application can signal misbehaviour with a negative score or gate
  peers before an application specific handshake is completed.
- P₆: IP colocation factor. This parameter is a threshold for the number of peers using the same IP
  address. If the number of peers in the same IP exceeds the threshold, then the value is the square
  of the surpluss, otherwise it is 0. This is intended to make it difficult to carry out sybil attacks
  by using a small number of IPs. The parameter is mixed with a negative weight.

> ### Topic Parameter Calculation and Decay

Reviewer: ```suggestion
#### Topic Parameter Calculation and Decay
```

The topic parameters are implemented using counters maintained internally by the router
whenever an event of interest occurs. The counters _decay_ periodically so that their values are
not continuously increasing and ensure that a large positive or negative score isn't sticky for
the lifetime of the peer.

> #### P₁: Time in Mesh

Reviewer: ```suggestion
##### P₁: Time in Mesh
```

In order to compute P₁, the router records the time when the peer is GRAFTed. The time in mesh
is calculated lazily during the decay update to avoid a large number of calls to `gettimeofday`.
The parameter value is the division of the time ellapsed since the GRAFT with an application
configurable quantum.

In pseudo-go:
```go
// topic configuration parameters
var TimeInMeshQuantum time.Duration
var TimeInMeshCap     float64

// lazily updated time in mesh
var meshTime time.Duration

// P₁
p1 := float64(meshTime / TimeInMeshQuantum)
if p1 > TimeInMeshCap {
  p1 = TimeInMeshCap
}
```

> #### P₂: First Message Deliveries

Reviewer: ```suggestion
##### P₂: First Message Deliveries
```

In order to compute P₂, the router maintains a counter that increments whenever a message
is first delivered in the topic by the peer. The parameter has a cap that applies at the time
of increment.

In pseudo-go:
```go
// topic configuration parameters
var FirstMessageDeliveriesCap float64

// couner updated every time a first message delivery occurs
var firstMessageDeliveries float64

// counter update
firstMessageDeliveries += 1
if firstMessageDeliveries > FirstMessageDeliveriesCap {
  firstMessageDeliveries = FirstMessageDeliveriesCap
}

// P₂
p2 := firstMessageDeliveries
```

> #### P₃ and P₃b: Mesh Message Delivery

Reviewer: ```suggestion
##### P₃ and P₃b: Mesh Message Delivery
```

var MeshMessageDeliveriesWindow, MeshMessageDeliveriesActivation time.Duration

// time in mesh, lazily updated
var meshTime time.Duration

// counter updated every time a first or near-first message delivery occurs by a mesh peer
var meshMessageDeliveries float64

// counter update
meshMessageDeliveries += 1
if meshMessageDeliveries > MeshMessageDeliveriesCap {
  meshMessageDeliveries = MeshMessageDeliveriesCap
}

// calculation of P₃
var deficit float64
if meshTime > MeshMessageDeliveriesActivation && meshMessageDeliveries < MeshMessageDeliveriesThreshold {
  deficit = MeshMessageDeliveriesThreshold - meshMessageDeliveries
}

p3 := deficit * deficit
```

In order to calculate P₃b, the router maintains a counter that is updated whenever the peer is pruned
with an active deficit in message delivery. The parameter is uncapped.

In pseudo-go:
```go
// counter updated at prune time
var meshFailurePenalty float64

// counter update
if meshTime > MeshMessageDeliveriesActivation && meshMessageDeliveries < MeshMessageDeliveriesThreshold {
  deficit = MeshMessageDeliveriesThreshold - meshMessageDeliveries
  meshFailurePenalty += deficit * deficit
}

// P₃b
p3b := meshFailurePenalty
```

> #### P₄: Invalid Messages

Reviewer: ```suggestion
##### P₄: Invalid Messages
```

In order to compute P₄, the router maintains a counter that increments whenever a message fails
validation. The counter is uncapped.

In pseudo-go:
```go
// counter updated every time a message fails validation
var invalidMessageDeliveries float64

// counter update
invalidMessageDeliveries += 1

// P₄
p4 := invalidMessageDeliveries
```

> #### Parameter Decay

Reviewer: ```suggestion
##### Parameter Decay
```

The counters associated with P₂, P₃, P₃b, and P₄ decay periodically by multiplying with a configurable
decay factor. When the value drops below a threshold it is considered zero.

In pseudo-go:
```go
// decay factors
var FirstMessageDeliveriesDecay, MeshMessageDeliveriesDecay, MeshFailurePenaltyDecay, InvalidMessageDeliveriesDecay float64

// 0-threshold
var DecayToZero float64

// periodic decay of counters
firstMessageDeliveries *= FirstMessageDeliveriesDecay
if firstMessageDeliveries < DecayToZero {
  firstMessageDeliveries = 0
}

meshMessageDeliveries *= MeshMessageDeliveriesDecay
if meshMessageDeliveries < DecayToZero {
  meshMessageDeliveries = 0
}

meshFailurePenalty *= MeshFailurePenaltyDecay
if meshFailurePenalty < DecayToZero {
  meshFailurePenalty = 0
}

invalidMessageDeliveries *= InvalidMessageDeliveriesDecay
if invalidMessageDeliveries < DecayToZero {
  invalidMessageDeliveries = 0
}
```

> ### Guidelines for Tuning the Scoring Function

Reviewer: ```suggestion
#### Guidelines for Tuning the Scoring Function
```

TBD

> ## Spam Protection Measures

Reviewer: ```suggestion
### Spam Protection Measures
```

    - [The Score Function](#the-score-function)
    - [Topic Parameter Calculation and Decay](#topic-parameter-calculation-and-decay)
      - [P₁: Time in Mesh](#p%E2%82%81-time-in-mesh)
      - [P₂: First Message Deliveries](#p%E2%82%82-first-message-deliveries)
      - [P₃ and P₃b: Mesh Message Delivery](#p%E2%82%83-and-p%E2%82%83b-mesh-message-delivery)
      - [P₄: Invalid Messages](#p%E2%82%84-invalid-messages)
      - [Parameter Decay](#parameter-decay)
    - [Guidelines for Tuning the Scoring Function](#guidelines-for-tuning-the-scoring-function)
  - [Spam Protection Measures](#spam-protection-measures)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

This document specifies extensions to [gossipsub v1.0](gossipsub-v1.0.md) intended to improve
bootstrapping and protocol attack resistant. The extensions change the algorithms that
prescribe local peer behaviour and are fully backwards compatible with v1.0 of the protocol.
Peers that implement these extensions, advertise v1.1 of the protocol using `/meshsub/1.1.0`
as the protocol string.

## Attack Vectors considered

`To be written`

## Protocol extensions 

> ### Peer Exchange

Reviewer: ```suggestion
### Peer Exchange on PRUNE
```

Gossipsub relies on ambient peer discovery in order to find peers within a topic of interest.
This puts pressure to the implementation of a scalable peer discovery service that
can support the protocol. With Peer Exchange, the protocol can now bootstrap from a small
set of nodes, without relying on an external peer discovery service.

Peer Exchange (PX) kicks in when pruning a mesh because of oversubscription. Instead of simply
telling the pruned peer to go away, the pruning peer _may_ provide a set of other peers where the
pruned peer can connect to reform its mesh. In addition, both the pruned and the pruning peer
add a backoff period from each other, within which they will not try to regraft. Both the pruning
and the pruned peer will immediate prune a `GRAFT` within the backoff period.
The recommended duration for the back period is 1 minute.

> 

Author: We can make it configurable.

> 

Author: It is configurable in the implementation, this is just the recommended value.

In order to implement PX, we extend the `PRUNE` control message to include an optional set of

> peers the pruned peer can connect to. This set of includes the Peer ID and a [_signed_ peer

Author: sure.

record](https://github.com/libp2p/specs/pull/217) for each peer exchanged.
In order to facilitate transion to the usage of signed peer records within the libp2p ecosystem,
the emitting peer is allowed to omit the signed peer record if it doesn't have one.
In this case, the pruned peer will have to utilize an external service to discover addresses for

> the peer, eg the DHT.

Author: sure, isn't that clear?

#### Protobuf

The `ControlPrune` message is extended with a `peer` field as follows.

```protobuf
message ControlPrune {
	optional string topicID = 1;

> 	repeated PeerInfo peers = 2; // gossipsub v1.1 PX

Author: It's configurable, in the implementation I set to `D` so it sends info for enough peers to form a full mesh.
Probably worth mentioning something.

> 	repeated PeerInfo peers = 2; // gossipsub v1.1 PX

Author: But not here, this is the protobuf; the 2 is irrelevant, it's the field ID in protobuf.

> 	repeated PeerInfo peers = 2; // gossipsub v1.1 PX

Author: I added a sentence in the PX paragraph that the recommended number of peers is `D`.

}

message PeerInfo {
	optional bytes peerID = 1;

> 	optional bytes signedPeerRecord = 2;

Reviewer: @vyzo does this mean that the Peer sending the PeerInfos, needs to request to every other peer an up to date SignedPeerRecord to be sent with the response? If yes, then I don't think the peer that is issuing the PRUNE will have the time to wait (as it might be overloaded with messages)

Author: Not at all. You send the signed peer records if you have them, otherwise you just send the peer ID and the pruned peer can find the peer in the DHT.

Reviewer: Ack. Thanks!

}

```

> ### Flood Publishing

Reviewer: @vyzo this extension reads more as a consideration than a specification. 

When should a Peer decide to use mesh/fanout or flood? Is there a randomizer for the strategy used?

Author: It's an application specification option. The recommendation is that it's used, but the application may turn it off if running in a controlled environment.

This behaviour also reduces message propagation latency as the message is injected to more points
in the network.

### Adaptive Gossip Dissemination

In gossipsub v1.0 gossip is emitted to a fixed number of peers, as specified by the `D_lazy`
parameter. In gossipsub v1.1 the disemmination of gossip is adaptive; instead of emitting gossip
to a fixed number of peers, we emit gossip to a percentage of our peers with a minimum of `D_lazy`
peers.

The parameter controlling the emission of gossip is called the gossip _factor_. When a node wants
to emit gossip during the heartbeat, first it selects all peers with a peer score above a gossip
threshold (see Peer Scoring below). From these peers, it randomly selects gossip factor peers with
a minimum of `D_lazy`, and emits gossip to the selected peers.

The recommended value for the gossip factor is `0.25`, which with the default of 3 rounds of gossip
per message ensures that each peer has at least 50% chance of receiving gossip about a message.
More specifically, for 3 rounds of gossip, the probability of a peer _not_ receiving gossip about
a fresh message is `(3/4)³=27/64=0.421875`. So each peer receives gossip about a fresh message with

> a `0.578125` probability.

Author: No, it doesn't. This is independent probability choice; each peer has a 25% chance of being selected for gossip for each round --- 3 rounds and the probability is 57%.

and is not shared. The score is a real value, computed as weighted mix of parameters,
with pluggable application specific scoring. The score is computed across all (configured) topics
with a weighted mix, such that faulty behaviour in one topic percolates to other topics.
Furthermore, the score is retained for some period of time when a peer disconnects, so that malicious
peers cannot easily reset their score when it drops to negative and well behaving
peers don't lose their status because of a disconnection.

The intention is to detect malicious or faulty behaviour and penalize the misbehaving peers
with a negative score.

#### Score Thresholds

The score is plugged into various gossipsub algorithms such that peers with negative scores are
removed from the mesh. Peers with heavily negative score are further penalized or even ignored
if the score drops too low.

More specifically, the following thresholds apply:
- `0`: the baseline threshold; peers with a score below this threshold are pruned from the mesh
  during the heartbeat and ignored when looking for peers to graft. Furthermore, no PX information

>   is emitted towards those peers and PX is ignored from them.

Author: It is mentioned below with the gossip threshold.

- `gossipThreshold`: when a peer's score drops below this threshold, no gossip is emitted towards
  that peer and gossip from that peer is ignored. This threshold should be negative, such that
  some information can be propagated to/from mildly negatively scoring peers.
- `publishThreshold`: when a peer's score drops below this threshold, self published messages are
  not propagated towards this peer when (flood) publishing. This threshold should be negative, and

>   less than or equal to the gossip threshold.

Author: They can publish all they want, we just don't publish to them if their score drops below `publishThreshold`.

#### The Score Function

The score function is a weighted mix of parameters, 4 of them per topic and 2 of them globally
applicable.
```
Score(p) = Σtᵢ*(w₁(tᵢ)*P₁(tᵢ) + w₂(tᵢ)*P₂(tᵢ) + w₃(tᵢ)*P₃(tᵢ) + w₃b(tᵢ)*P₃b(tᵢ) + w₄(tᵢ)*P₄(tᵢ)) + w₅*P₅ + w₆*P₆
```
where `tᵢ` is the topic weight for each topic where per topic parameters apply.

The parameters are defined as follows:
- P₁: time in mesh for a topic. This is the time a peer has been in the mesh, capped to a small value
  and mixed with a small positive weight. This is intended to boost peers already in the mesh so that
  they are not prematurely pruned because of oversubscription.
- P₂: first message deliveries for a topic. This is the number of message first delivered by the peer
  in the topic, mixed with a positive weight. This is intended to reward peers who first forward a
  valid message.
- P₃: mesh message delivery rate for a topic. This parameter is a threshold for the expected message
  delivery rate within the mesh in the topic. If the number of deliveries is above the threshold,
  then the value is 0. If the number is below the threshold, then the value of the parameter is

>   the square of the deficit.

Author: The threshold is application configured, not automatically computed.

  mesh. The parameter is mixed with negative weight.
- P₄: invalid messages for a topic. This is he number of invalid messages delivered in the topic.
  This is intended to penalize peers who transmit invalid messages, according to application specific
  validation rules. It is mixed with a negative weight.
- P₅: application specific score. This is the score component assigned to the peer by the application
  itself, using application specific rules. The weight is positive, but the parameter itself has an
  arbitrary real value, so that the application can signal misbehaviour with a negative score or gate
  peers before an application specific handshake is completed.
- P₆: IP colocation factor. This parameter is a threshold for the number of peers using the same IP
  address. If the number of peers in the same IP exceeds the threshold, then the value is the square
  of the surpluss, otherwise it is 0. This is intended to make it difficult to carry out sybil attacks
  by using a small number of IPs. The parameter is mixed with a negative weight.

#### Topic Parameter Calculation and Decay

The topic parameters are implemented using counters maintained internally by the router

> whenever an event of interest occurs. The counters _decay_ periodically so that their values are

Reviewer: What is "periodically" and by how much? Is this a parameter? What are the defaults?

Author: It is an application parameter. 1-10s is a good value.

Author: I will add something about it in the guidelines.

In pseudo-go:
```go
// counter updated every time a message fails validation
var invalidMessageDeliveries float64

// counter update
invalidMessageDeliveries += 1

// P₄
p4 := invalidMessageDeliveries
```

##### Parameter Decay

The counters associated with P₂, P₃, P₃b, and P₄ decay periodically by multiplying with a configurable
decay factor. When the value drops below a threshold it is considered zero.

In pseudo-go:
```go
// decay factors
var FirstMessageDeliveriesDecay, MeshMessageDeliveriesDecay, MeshFailurePenaltyDecay, InvalidMessageDeliveriesDecay float64

// 0-threshold
var DecayToZero float64

// periodic decay of counters
firstMessageDeliveries *= FirstMessageDeliveriesDecay
if firstMessageDeliveries < DecayToZero {
  firstMessageDeliveries = 0
}

meshMessageDeliveries *= MeshMessageDeliveriesDecay
if meshMessageDeliveries < DecayToZero {
  meshMessageDeliveries = 0
}

meshFailurePenalty *= MeshFailurePenaltyDecay
if meshFailurePenalty < DecayToZero {
  meshFailurePenalty = 0
}

invalidMessageDeliveries *= InvalidMessageDeliveriesDecay
if invalidMessageDeliveries < DecayToZero {
  invalidMessageDeliveries = 0
}
```

#### Guidelines for Tuning the Scoring Function

> TBD

Reviewer: @vyzo any ETA?

Author: once we've made some progress with the testground tests.

This puts pressure to the implementation of a scalable peer discovery service that
can support the protocol. With Peer Exchange, the protocol can now bootstrap from a small
set of nodes, without relying on an external peer discovery service.

Peer Exchange (PX) kicks in when pruning a mesh because of oversubscription. Instead of simply
telling the pruned peer to go away, the pruning peer _may_ provide a set of other peers where the
pruned peer can connect to reform its mesh (see Peer Scoring below).
In addition, both the pruned and the pruning peer add a backoff period from each other, within which
they will not try to regraft. Both the pruning and the pruned peer will immediate prune a `GRAFT`
within the backoff period.
The recommended duration for the backoff period is 1 minute, while the recommended number of peers
to exchange is equal to `D` so that the pruned peer can form a full mesh.

In order to implement PX, we extend the `PRUNE` control message to include an optional set of
peers the pruned peer can connect to. This set of peers includes the Peer ID and a [_signed_ peer
record](https://github.com/libp2p/specs/pull/217) for each peer exchanged.

> In order to facilitate transion to the usage of signed peer records within the libp2p ecosystem,

Reviewer 3: ```suggestion
In order to facilitate the transition to the usage of signed peer records within the libp2p ecosystem,
```
